halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/plugins/Plugin.java

```diff
@@ -72,4 +90,86 @@ public InputStream downloadManifest() {
               .build());
     }
   }
+
+  /**
+   * Normalizes plugin options by converting YAML keys with dot notation into a nested HashMap
+   *
+   * @param opts
+   * @return
+   */
+  private Map<String, Object> normalizeOptions(Map<String, Object> opts) {
+    return opts.entrySet().stream()
+        .collect(
+            Collector.of(
+                HashMap::new,
+                (left, entry) -> left.putAll(parseOption(entry.getKey(), entry.getValue())),
+                (left, entry) -> left));
+  }
+
+  /**
+   * Used to merge plugin options passed in through the hal config to overwrite the default plugin
+   * options, as necessary. Since options can be a nested map, if a hal config overwrites a
+   * particular piece of the default plugin options, we want to preserve the other options.
+   *
+   * <p>If a key is present on both the original and new options, the types match, and are of type
+   * Map or List, then we merge the two values. Otherwise, we overwrite the keys with the values in
+   * the newMap.
+   *
+   * @param original
+   * @param newMap
+   * @return
+   */
+  public static Map<String, Object> merge(
+      Map<String, Object> original, Map<String, Object> newMap) {
+    for (String key : newMap.keySet()) {
+      if (newMap.get(key) instanceof Map && original.get(key) instanceof Map) {
+        Map originalChild = (Map) original.get(key);
+        Map newChild = (Map) newMap.get(key);
+        original.put(key, Plugin.merge(originalChild, newChild));
+      } else if (newMap.get(key) instanceof List && original.get(key) instanceof List) {
+        List originalChild = (List) original.get(key);
+        List newChild = (List) newMap.get(key);
+        for (Object each : newChild) {
+          if (!originalChild.contains(each)) {
+            originalChild.add(each);
```

Comment:
This is `N^2` in the list size; not sure it matters given what I'd imagine to be typical list sizes in config files, but worth keeping in mind.